docs/modules/password_auth_provider_callbacks.md

```diff
@@ -105,6 +105,64 @@ device ID), and the (now deactivated) access token.
 
 If multiple modules implement this callback, Synapse runs them all in order.
 
+### `get_username_for_registration`
+
+_First introduced in Synapse v1.52.0_
+
+```python
+async def get_username_for_registration(
+    uia_results: Dict[str, Any],
+    params: Dict[str, Any],
+) -> Optional[str]
+```
+
+Called when registering a new user, allowing the module to define the username to give the user.
```

Comment:
I don't think we actually spell out anywhere that the return value the username the module wants to be used. Maybe something like this?

Called when registering a new user. Modules can choose a username for a user by returning the desired username from this callback.